### PR TITLE
implement a very basic TLS 1.3 handshake

### PR DESCRIPTION
**Pull request by marten-seemann**

### CODE REVIEW

File: crypto/aesgcm_aead.go [hunk 1]
```diff
@@ -17,31 +16,34 @@ type aeadAESGCM struct {
...
+// NewAEADAESGCM creates a AEAD using AES-GCM
 func NewAEADAESGCM(otherKey []byte, myKey []byte, otherIV []byte, myIV []byte) (AEAD, error) {
-	if len(myKey) != 16 || len(otherKey) != 16 || len(myIV) != 4 || len(otherIV) != 4 {
-		return nil, errors.New("AES-GCM: expected 16-byte keys and 4-byte IVs")
+	// the IVs need to be at least 8 bytes long, otherwise we can't compute the nonce
+	if len(otherIV) != ivLen || len(myIV) != ivLen {
+		return nil, errors.New("AES-GCM: expected 12 byte IVs")
 	}
-	encrypterCipher, err := aes12.NewCipher(myKey)
+
```
Comment (lucas-clemente) [inline crypto/aesgcm_aead.go]: Is there a reason why you're not checking the key size here (anymore)?

File: crypto/key_derivation.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	"github.com/bifurcation/mint"
+	"github.com/lucas-clemente/quic-go/internal/protocol"
+)
+
+const (
+	clientExporterLabel = "EXPORTER-QUIC client 1-RTT Secret"
+	serverExporterLabel = "EXPORTER-QUIC server 1-RTT Secret"
+)
+
+// MintState contains
```
Comment (lucas-clemente) [inline crypto/key_derivation.go]: Nit

File: crypto/key_derivation.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	}
+	otherKey, otherIV, err := computeKeyAndIV(ms, otherLabel)
+	if err != nil {
+		return nil, err
+	}
+	return NewAEADAESGCM(otherKey, myKey, otherIV, myIV)
+}
+
+func computeKeyAndIV(ms MintState, label string) (key, iv []byte, err error) {
+	cs := ms.GetCipherSuite()
```
Comment (lucas-clemente) [inline crypto/key_derivation.go]: Should we check here that this is actually AES-GCM?

File: handshake/crypto_setup_tls.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	if perspective == protocol.PerspectiveServer {
+		conn = mint.Server(&fakeConn{cryptoStream}, mintConf)
+	} else {
+		conn = mint.Client(&fakeConn{cryptoStream}, mintConf)
+	}
+	return &cryptoSetupServerTLS{
+		perspective:   perspective,
+		mintConf:      mintConf,
+		conn:          conn,
+		nullAEAD:      crypto.NewNullAEAD(perspective, protocol.Version39),
```
Comment (lucas-clemente) [inline handshake/crypto_setup_tls.go]: Why is there a version const here?

File: handshake/crypto_setup_tls.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+		keyDerivation: crypto.DeriveAESKeys,
+		aeadChanged:   aeadChanged,
+	}, nil
+}
+
+func (h *cryptoSetupServerTLS) HandleCryptoStream() error {
+	alert := h.conn.Handshake()
+	utils.Debugf("ALERT: %#v", alert)
```
Comment (lucas-clemente) [inline handshake/crypto_setup_tls.go]: Remove this line or move it into the if block below.

File: crypto/aesgcm_aead.go (hunk 1 above)
Comment (marten-seemann) [reply crypto/aesgcm_aead.go]: Yes, `aes.NewCipher` can be called with different key sizes, so it performs the check if the key has one of the accepted sizes. See https://golang.org/pkg/crypto/aes/#NewCipher

Comment (lucas-clemente) [reply crypto/aesgcm_aead.go]: Okay, and mint will check for us that we don't accept weak keys?

Comment (marten-seemann) [reply crypto/aesgcm_aead.go]: TLS 1.3 doesn't have any weak cipher suites. These are the ones supported by mint: https://github.com/bifurcation/mint/blob/master/common.go#L43-L52.

File: crypto/key_derivation.go (hunk 3 above)
Comment (marten-seemann) [reply crypto/key_derivation.go]: We probably should do that. That will mean checking the cipher suite against the values defined in https://github.com/bifurcation/mint/blob/master/common.go#L43-L52 though, I don't see a general way to check for AES.

File: handshake/crypto_setup_tls.go (hunk 4 above)
Comment (marten-seemann) [reply handshake/crypto_setup_tls.go]: I think I forgot to write the correct null AEAD constructor. I will add this later today.

File: crypto/key_derivation.go (hunk 3 above)
Comment (marten-seemann) [reply crypto/key_derivation.go]: Actually, what we should do is set a list of supported ciphers in the `mint.Config`, and only use AES ciphers there. We can then trust mint to negotiate one of them.

File: handshake/crypto_setup_tls.go (hunk 4 above)
Comment (marten-seemann) [reply handshake/crypto_setup_tls.go]: Fixed.

File: crypto/aesgcm_aead.go (hunk 1 above)
Comment (marten-seemann) [reply crypto/aesgcm_aead.go]: I added tests with 16 and 32 byte keys, and another test that we error on invalid key sizes.

**Comment by marten-seemann** [comment]
@lucas-clemente: I think I addressed all your comments. Furthermore, I moved all mint-related methods into one interface, which should make testing easier. PHAL.

**Pull request merged**